tutorial/decorators.md

```diff
@@ -0,0 +1,243 @@
+
+An example shows how pyccel handles negative indexes beween Python and C:
+
+```python
+from pyccel.decorators import allow_negative_index
+
+@allow_negative_index('a')
+def fun1():
+    a = [1,2,3,4,5,6]
+    print(a[-1])
+
+    b = [1,2,3,4,5,6]
+    print(b[-1])
+```
+
+This is the generated C code:
+
+```C
+
+#include "boo.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include "ndarrays.h"
+
+
+/*........................................*/
+void fun1(void)
+{
+    t_ndarray a;
+    t_ndarray b;
+    a = array_create(1, (int64_t[]){6}, nd_int64);
+    int64_t array_dummy_0001[] = {1, 2, 3, 4, 5, 6};
+    memcpy(a.nd_int64, array_dummy_0001, a.buffer_size);
+    printf("%ld\n", GET_ELEMENT(a, nd_int64, 5));
+    b = array_create(1, (int64_t[]){6}, nd_int64);
+    int64_t array_dummy_0002[] = {1, 2, 3, 4, 5, 6};
+    memcpy(b.nd_int64, array_dummy_0002, b.buffer_size);
+    printf("%ld\n", GET_ELEMENT(b, nd_int64, 5));
+}
+/*........................................*/
+```
+
+And here is the equivalent Fortran code:
+
+```fortran
+module boo
+
+
+  use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T
+  implicit none
+
+  contains
+
+  !........................................
+  subroutine fun1()
+
+    implicit none
+
+    integer(i64), allocatable :: a(:)
+    integer(i64), allocatable :: b(:)
+
+    allocate(a(0:5_i64))
+    a = [1_i64, 2_i64, 3_i64, 4_i64, 5_i64, 6_i64]
+    print *, a(size(a, kind=i64) - 1_i64)
+    allocate(b(0:5_i64))
+    b = [1_i64, 2_i64, 3_i64, 4_i64, 5_i64, 6_i64]
+    print *, b(size(b, kind=i64) - 1_i64)
+
+  end subroutine fun1
+  !........................................
+
+end module boo
+```
+
+## Elemental and pure decorators
+
+This decorator indicates that the function bellow is an elemental one, An elemental function is a function with a single scalar operator and a scalar return value which can also be called on an array. When it is called on an array it returns the result of the function called elementwise on the array.
+This decorator `pure` indicates that the function bellow is a pure one. So that function should return identical return values for identical arguments and has no side effects in its application.
+
+Here is a simple usage example:
+
+```python
+from pyccel.decorators import elemental, pure
+
+@pure
+@elemental
+@types(float)
+def square(x):
+    s = x*x
+    return s
+```
+
+This is the C generated code, this code can be generated even without these two decorators due to the C language itself that has not the elemenwise feature in the manipulation of the arrays nor a function prefix `pure`. So these decorators will not make a big deffirence in the Python/C conversion:
```

Comment:
Typo `deffirence`->`difference`

The `pure` decorator does nothing in c (which does not have an equivalent keyword), but `elemental` does. You won't see it in the translation of the function itself, but in the translation of a call. E.g:
```
a = np.ones(5)
square(a)
```
Without `@elemental` for the function `square` this will not compile. It should raise a semantic error. With `@elemental` it will be translated to a for loop

Reply:
Typo `elemenwise` -> `elementwise` (missing `t`)